rfcs/20250421-source-target-pairs.md

```diff
@@ -0,0 +1,136 @@
+# [RFC] Add source-target pairs to send/recv ops
+
+Status: In Review<br/>
+Initial version: 04/21/2025<br/>
+Last updated: 04/21/2025<br/>
+Discussion thread: N/A
+
+## Overview
+
+This RFC proposes adding a new attribute `source_target_pairs` to `send` and
+`recv` ops. `source_target_pairs` allows users to specify peer-to-peer
+communication patterns using global device IDs (zero-indexed integers).
+Currently this feature is only available on GPUs.
+
+## Background
+
+SPMD-based pipeline parallelism relies on optimizations in XLA to pipeline
+send/recv operations in such a way that compute and communication are
+overlapped. The user expresses this through collective permutes and relies on
+XLA to decompose these into send/recv operations, which are then pipelined
```

Comment:
This is more for my understanding the motivation:

The hlo-level `send`/`recv`s, generated by the XLA decomposition, do carry the source/target info using attribs like ` frontend_attributes {_xla_send_recv_source_target_pairs={{3,0}}}`. I am not clear how exposing this information at a high-level Stablehlo representation would help as the only stablehlo operations, exported from Jax, are collective permutes (not the stablehlo send/recv's).

Reply:
Per the last sentence in this paragraph, I'm also working on adding `psend` and `precv` to `jax.lax.parallel`. I didn't include it in this RFC because I figured it's not related to stableHLO.

Reply:
@rosiezou Is there an available RFC for the JAX side of things.

Reply:
Not an official RFC but https://github.com/jax-ml/jax/pull/28101 is a WIP of the JAX changes.